Fix issue #34242: Attach/detach should recover from a crash

tsmetana opened the pull request:
When the attach/detach controller crashes and a pod with attached PV is deleted afterwards the controller will never detach the pod's attached volumes. To prevent this the controller should try to recover the state from the nodes status and figure out which volumes to detach. This requires some changes in the volume providers too: the only information available from the nodes is the volume name and the device path. The controller needs to find the correct volume plugin and reconstruct the volume spec just from the name. This required a small change also in the volume plugin interface.

Fixes Issue #34242.
cc: @jsafrane @jingxu97

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39732)
<!-- Reviewable:end -->

jsafrane commented:
Assigning to @jingxu97, she knows attach/detach logic better

tsmetana commented:
@k8s-bot gci gke e2e test this issue: #23634

gnufied commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -209,13 +209,62 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+	}
+
+	return nil
+}
```
I think we need some tests to verify if these two methods indeed reconstructs the actual and desired state of world correctly via API calls.

tsmetana replied:
Right. I only was not sure how to test the changes exactly: the biggest problem is that I had to change all the volume plugins so ideally they should be all tested. But for this part I think I should be able to come with something. Thanks.

kerneltime commented:
Are there any automated tests for this ? If not what testing was done? 

Looking into the code change but from a first glance the reconstruction of volume spec might need fixing for vSphere.

kerneltime commented on pkg/volume/vsphere_volume/vsphere_volume.go:
```diff
@@ -120,6 +120,18 @@ func (plugin *vsphereVolumePlugin) newUnmounterInternal(volName string, podUID t
+		Name: volumeName,
+		VolumeSource: v1.VolumeSource{
+			VsphereVolume: &v1.VsphereVirtualDiskVolumeSource{
+				VolumePath: volumeName,
```
This will need fixing.. @BaluDontu

tsmetana replied:
OK. Thank for the response. I tested the patch in AWS and the purpose of the PR was mainly to solicit feedback on the patch since it's actually my first actual contribution to kubernetes.

kerneltime replied:
Let me know if you need us to address this code if you plan to merge.

tsmetana replied:
Thanks @kerneltime. I've looked at the plugins code and I really filled the spec with nonsense. Probably at more places than just vsphere... I would definitely appreciate any help: I'm not able to test all the plugins myself: I'm quite sure those that don't need mountPath to reconstruct volumeSpec are OK but I might need some guidance in the other cases.

tsmetana commented:
Sorry or the failed builds. I need to rebase (and fix) the patches.

childsb commented:
/approve

jingxu97 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -210,13 +210,65 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+	return nil
+}
+
+func (adc *attachDetachController) populateDesiredStateOfWorld() error {
```
A problem here is the race condition between this population and nodeAdd. If a pod is added before the node is added to the desired state, the podAdd will fail.
One solution could be get the node list from kubeClient and add all the nodes to the desired state before adding pods

tsmetana replied:
And the funny part is I found this out myself and circumvented it in the unit tests (so they pass)... Will fix that.

jingxu97 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -210,13 +210,65 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+				glog.Errorf("could not construct volume spec in plugin %s for volume %s: %v", pluginName, volumeName, err)
+				continue
+			}
+			_, err = adc.actualStateOfWorld.AddVolumeNode(volumeSpec, nodeName, attachedVolume.DevicePath)
```
One thing related to AddVolumeNode() function is by default it set mountedByNodeSetCount: 0. Although it should cause any dangerous thing happen, but based on current logic, it might delay for reconciler to detach. I am ok with this but might worth to put a note mention this. Or you can get VolumesInUse and update this field.

tsmetana replied:
This is actually what I saw in the testing: the volume gets detached after some timeout (6 minutes I think). And after discussion with @jsafrane this should be addressed too.

jingxu97 replied:
Could you use Capital case for the first letter of error message to be consistent?

jingxu97 commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -189,6 +189,18 @@ func getVolumeSource(
+		Name: volName,
+		VolumeSource: v1.VolumeSource{
+			AWSElasticBlockStore: &v1.AWSElasticBlockStoreVolumeSource{
+				VolumeID: volName,
```
The volName is not the VolumeID. Maybe put an note about this.

jingxu97 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -210,13 +210,65 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+				glog.Errorf("could not find plugin named %s for volume %s: %v", pluginName, volumeName, err)
+				continue
+			}
+			volumeSpec, err := plug.ConstructVolumeSpecFromName(volumeName)
```
ConstructVolumeSpecFromName function fake volumeSource information for almost all of volume plugins. The main issue is that this information will be used by VerifyVolumesAreAttached(). If this information (e.g., volumeID) is wrong, the function will return wrong information related whether volumes are attached or not. 

One way to avoid this is populate desired state first which contains this information, and then find the matching volume to populate this spec information from desired state. But volume spec information might not be recovered if pod is deleted so desired state does not have it either. PR #40239 might help solve this issue, but this PR is revered. Will follow up with the status.

tsmetana replied:
Yes. The ConstructVolumeSpecFromName fills the spec with wrong data. It's been noted above in the VSphere plugin. This is the most problematic part of the patch since I really had no idea how to fix that properly. I'll take a look a the other PR. (Or if somebody came with an idea I'm one ear.) Thanks.

wongma7 replied:
It looks to me like PR #40239 would remove the need for recovery in the case where a crash happens before unmounting has succeeded. Because then the pod wouldn't be allowed to be deleted in the first place, and the controller could find it and populate its states as usual.

But if a crash happens after unmounting has already succeeded & before detach happens, then recovery is still needed because the pod would be allowed to be deleted & won't be around to inspect.

So if names from node.Status are not enough for reconstruction then you may have to serialize & store entire volume specs somewhere? Maybe as ugly annotations on the node? I have no idea what the "proper" way to do it is.

tsmetana replied:
I went through the plugins code wondering why the volumes get detached in my tests since I really don't put the correct data to the spec. Seems it's actually by accident: I use the unique volume name from the node and split it into the plugin name and volume name. The volume name I incidentally ends with the same component (path-wise) as the mount path. The Detach() plugin methods usually start with extracting the last component of their first argument. If this is the volume name I have the result is identical as it would be with the correct volume source and the detach operation succeeds (@kerneltime -- I think this is the same also in the vSphere case).
The question is whether I should document this in the code and rely on this behaviour or some deeper re-design of the AttachDetach controller would be required.

jingxu97 replied:
Sorry for the confusion about the names. You are right. UniqueVolume is constructed as pluginName/volumeSourceName. This  volumeSourceName is from plugin.GetVolumeName() which takes the source.volumeID (or other names) from different plugins. So when you ConstructVolumeSpecFromName() as below, the volumeID is actually the volumeName you extract from SplitUniqueName(). But the meta Name: volumeName for the Spec is wrong. I checked our code and I didn't find anywhere this meta name is actually used. So, overall your approach should work.

tsmetana replied:
Thanks. I'm actually trying one more thing: store the string that I'm unable to generate from the volume name also in the node status. This would be more robust, however it's an API change (small one though).

jingxu97 replied:
Can you provide more details about your plan on API change? I actually don't recommend to have API change in this PR unless it is really necessary because it is a very complicated and long process to get approved.

tsmetana replied:
Right now the node status looks like this:
```
status:
    volumesAttached:
    - devicePath: /dev/vdb
      name: kubernetes.io/...
```
So I was thinking of adding one more string there that would contain the part I'm unable to reconstruct from the name (e.g., PDName in case of gce_pd):
```
status:
    volumesAttached:
    - devicePath: /dev/vdb
      name: kubernetes.io/plugin/volume_name
      detachVolumeName: whatever_the_plugin_needs
```
I talked to @jsafrane today and he also thinks getting the API change approval would be quite a complicated thing and suggested to store this data in the node annotations. Which I suppose would work too...

jingxu97 commented:
```
status:
    volumesAttached:
    - devicePath: /dev/vdb
      name: kubernetes.io/plugin/volume_name
      detachVolumeName: whatever_the_plugin_needs
```
I believe the above detachVolumeName is not really used anywhere in volume controller. At least for this PR, I am ok with the fake name.

jingxu97 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,77 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 
+	err := adc.populateActualStateOfWorld()
+	if err != nil {
+		glog.Errorf("error populating the actual state of world: %v", err)
```
Could you use capital case for the first letter to be consistent?

jingxu97 commented:
@tsmetana do you plan to finish this PR soon? Overall this PR is good to me. The only thing I suggest is you might want to update the volume for mountedByNodeSetCount. This information is basically describing whether the volume is mounted by node or not. This information can be retrieved by node status volumeInUse field. You can use nodeStatusUpdater to query this information from API server. Please let me know if you have any questions about it.

tsmetana commented:
@jingxu97 Yes, I would really like to have this finished soon... It takes much longer than I had expected. I have an improved (hopefully) version that fixes SetVolumeMountedByNode so the volume is marked as unmounted when not in use. I'm also trying store the plugin-specific part of the volume spec in the node annotation so eventually the plugin interface nor the plugin need to be changed. I really hope to have this done by the end of this week.

tsmetana commented:
Squashed and force-pushed new version of the patch: it contains the fix for SetVolumeMountedByNode method and saves the plugin-specific part of the volume spec in node annotations which I believe is more robust than the original solution.
Any comments and suggestions are welcome.

tsmetana commented:
Please hold on with reviewing/commenting: after diving into the operation_executor code I realized that if the volume unique name is really plugin name + volume name the operation_executor is able to detach the volume without the spec... Most of the problematic parts of the patch are about the volume spec reconstruction which might be redundant and in that case I can simplify the patch. Sorry for the noise...

tsmetana commented:
Here's the updated fix that doesn't rely on changes in the plugin volumes since the spec is not required for detaching: I still rely on the fact the unique name consist of plugin name + volume name (which is all that's actually needed). I have also added a code that reconstruct the volume specs for all the volumes "known" to the pods (e.g. those that should not be detached) so the DesiredStateOfWorld and ActualStateOfWolrd are consistent for all the attached volumes after the recovery.

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -233,39 +234,61 @@ func (og *operationGenerator) GenerateAttachVolumeFunc(
 }
 
+func (og *operationGenerator) splitUniqueName(uniqueName v1.UniqueVolumeName) (string, string, error) {
+	components := strings.SplitN(string(uniqueName), "/", 3)
```
Please document expected format of uniqueName and link it to `GetUniqueVolumeNameFromSpec`.

Also, the code here expects that all volume plugin names contain exactly one '/'. This is OK with current code, however comment in plugin interface definition at https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L74 should be more strict - from 'should use namespaced names such as "example.com/volume"' to 'must use namespaced names ... and must contain exactly one "/" character.'

tsmetana replied:
Thanks. I will add/fix the comments.

ncdc commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,115 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 
+	err := adc.populateActualStateOfWorld()
+	if err != nil {
+		glog.Errorf("Error populating the actual state of world: %v", err)
```
Use `runtime.HandleError` instead of `glog.Errorf` everywhere.

Errors need to start with a lowercase letter.

e.g. `runtime.HandleError(fmt.Errorf("error populating the actual state of world: %v", err))`

tsmetana replied:
All the rest of the attachdetach controller uses `glog.Errorf`. That's why I'm using it too. I'd rather stay consistent with the existing code and then eventually refactor all the error cases (including the old ones). I'd prefer to do that in a separate PR.

ncdc commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,115 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+func (adc *attachDetachController) populateDesiredStateOfWorld() error {
+	glog.V(5).Infof("Populating DesiredStateOfworld")
+
+	nodes, err := adc.kubeClient.Core().Nodes().List(metav1.ListOptions{})
```
We should maybe consider using a node informer?

cc @deads2k

ncdc replied:
Oh I see this is just at startup, hmm.

tsmetana replied:
Yes, this is on purpose. I need to gather everything before the reconciler starts so the "StateOfWorld" objects reflect the reality.

ethernetdan commented:
Removing PR from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -233,39 +234,66 @@ func (og *operationGenerator) GenerateAttachVolumeFunc(
+// This means the unique name will always be in the form of plugin_namespace/plugin/volume_name
+// see k8s.io/kubernetes/pkg/volume/plugins.go VolumePlugin interface description and
+// pkg/volume/util/volumehelper/volumehelper.go GetUniqueVolumeNameFromSpec that constructs the unique volume names.
+func (og *operationGenerator) splitUniqueName(uniqueName v1.UniqueVolumeName) (string, string, error) {
```
This is a utility function, might be good to put in a utility file.

tsmetana replied:
OK. Will move it to the volumehelper where it was in the original patch version.

calebamiles commented:
moving this PR to the 1.7 milestone @tsmetana  

**cc: @ethernetdan, @kubernetes/kubernetes-release-managers**

jsafrane commented:
This is a bugfix and I'd like to have it fixed in 1.6, adjusting milestone.

saad-ali commented:
Punting to post-1.6. This is too large a change to get in at this point.

tsmetana committed:
Fix issue #34242: Attach/detach should recover from a crash

When the attach/detach controller crashes and a pod with attached PV is deleted
afterwards the controller will never detach the pod's attached volumes. To
prevent this the controller should try to recover the state from the nodes
status.

tsmetana commented:
Here is the newer version of the patch: I've split the tests into a separate commit.

tsmetana commented:
The failed test is caused by an existing race in the attach/detach controller that the new tests trigger. It's unrelated to the patch. I've [opened a PR](https://github.com/kubernetes/kubernetes/pull/43289) that should fix that.

gnufied commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -298,7 +403,7 @@ func (adc *attachDetachController) nodeUpdate(oldObj, newObj interface{}) {
 		adc.desiredStateOfWorld.AddNode(nodeName)
 	}
-	adc.processVolumesInUse(nodeName, node.Status.VolumesInUse)
+	adc.processVolumesInUse(nodeName, node.Status.VolumesInUse, false)
```
can we write a comment next to `false` on the lines of `/*forceUnmount */`

saad-ali replied:
+1

gnufied commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -255,25 +255,34 @@ func (asw *actualStateOfWorld) AddVolumeToReportAsAttached(
+		// This happens only on controller startup when reading the volumes from node
+		// status; if the pods using the volume have been removed and are unreachable
+		// the volumes should be detached immediately and the spec is not needed
+		volumeName = uniqueName
```
So when `volumeSpec` is nil - we are adding `nil` spec to `asw.attachedVolumes` right? What I mean is:

```
volumeObj = attachedVolume{
			volumeName:      volumeName,
			spec:            volumeSpec,
			nodesAttachedTo: make(map[types.NodeName]nodeAttachedTo),
			devicePath:      devicePath,
		}
```

In this struct - spec will be clearly `nil`. Are you sure that this won't cause other parts of the code that don't expect spec to be nil? 

I see that you have handled some cases where `nil` spec might cause problems in operation_generator. But not all the cases. Take https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L123 code for example, which undeneath uses those specs to extract plugins to verify if volumes are attached.

I am thinking that - this will throw panic errors, no?

tsmetana replied:
Yes. This is the most dangerous part of the patch, I agree. The particular case you pointed out should be safe: the syncStates() method is being called only after the reconcile() and therefore there should exist no volumes with nil Spec in the ASW: the volume with nil spec is only present in ASW and therefore removed from the list of attached nodes immediately during the first reconcile() run -- syncStates() should never encounter it. But you made me to look at the tests again and I will add more nil pointer tests into them.

gnufied commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,117 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+				glog.Errorf("Error adding node to ActualStateOfWorld: %v", err)
+				continue
+			}
+			adc.processVolumesInUse(nodeName, nodes.Items[i].Status.VolumesInUse, true)
```
why is `forceUnmount` flag true here? Also - we should add the comment `true /*forceUnmount*/`

jingxu97 replied:
ping on this. And also add comments about the reason to set this flag to true

gnufied commented:
Overall this PR looks okay, my biggest concern is adding `nil` spec values in `asw.attachedVolumes` map. We may have to safeguard against it and make sure that `nil` reference doesn't causes crashes/panic errors.

tsmetana commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,117 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+	}
+
+	for i := range nodes.Items {
+		nodeName := types.NodeName(nodes.Items[i].Name)
```
Right. I still haven't got used to Go that well.

jingxu97 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,117 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+	volumeName v1.UniqueVolumeName, nodeName types.NodeName) (string, error) {
+	var devicePath string = ""
+	var found bool = false
+	node, err := adc.kubeClient.Core().Nodes().Get(string(nodeName), metav1.GetOptions{})
```
this API call is not necessary, devicePath retrieved earlier from populateActualStateOfWorld is already the correct one.

jingxu97 replied:
Another possible way to populateDesiredState first, then when populateActualState, we can update volume spec if it is in the desired state.

tsmetana replied:
This is being called when iterating over pods -- the devicePath is not needed for ActualStateOfWorld. I can however save some API calls by adding the nodes to DesiredStateOfWorld right away when examining the node statuses in ASW populator.

jingxu97 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,117 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+		for _, attachedVolume := range nodes.Items[i].Status.VolumesAttached {
+			uniqueName := attachedVolume.Name
+			glog.V(5).Infof("Adding to ASW: volume %s, node %s", uniqueName, nodeName)
+			err = adc.actualStateOfWorld.MarkVolumeAsAttached(uniqueName, nil, nodeName, attachedVolume.DevicePath)
```
nil pointer for volume spec will cause problem. Detach will use this information.

jingxu97 replied:
I remember in your previous implementation, you reconstruct the spec

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -314,35 +315,48 @@ func (og *operationGenerator) GenerateDetachVolumeFunc(
+		// Get attacher plugin and the volumeName by splitting the volume unique name in case
+		// there's no VolumeSpec: this happens only on attach/detach controller crash recovery
+		// when a pod has been deleted during the controller downtime
+		pluginName, volumeName, err = volumehelper.SplitUniqueName(volumeToDetach.VolumeName)
```
I see you now use uniqueName to get pluginName, but it seems not very safe to do this way. Although for current plulgins, it works but there is no guarantee the name will always construct like this. And what if there is / in the plugin name too?

jingxu97 replied:
What is the reason you change from reconstructSpec to this logic?

tsmetana replied:
I have always used the unique name to get the plugin name even in the previous version that was reconstructing the spec (the spec reconstruction was a plugin method so I had to have the plugin name first). Yes, this adds constraints on the plugin naming: it must follow the current pattern.

gnufied replied:
I talked with @jingxu97 bit offline and I think what we can do is - for now we should leave it - as it is but in future perhaps amend the node resource to store more information about volume, so as this hack need not be needed.

saad-ali replied:
Agreed pretty funky should be improved in the future

gnufied commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,117 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+			}
+			nodeName := types.NodeName(podToAdd.Spec.NodeName)
+			volumeName := v1.UniqueVolumeName(podVolume.Name)
+			if adc.actualStateOfWorld.VolumeNodeExists(volumeName, nodeName) {
```
@tsmetana ping on this. is `volumeName` being passed to `VolumeNodeExists` function accurate?

tsmetana replied:
@gnufied you're right. This is not the volume name I'm looking for. So this needs to be fixed too.

kargakis commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -230,13 +231,128 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+func (adc *attachDetachController) populateDesiredStateOfWorld() error {
+	glog.V(5).Infof("Populating DesiredStateOfworld")
+
+	pods, err := adc.kubeClient.Core().Pods(v1.NamespaceAll).List(metav1.ListOptions{})
```
This call bypasses the cache

kargakis commented on pkg/volume/util/volumehelper/volumehelper.go:
```diff
@@ -87,3 +88,18 @@ func GetUniqueVolumeNameFromSpec(
+// plugin_namespace/plugin/volume_name, see k8s.io/kubernetes/pkg/volume/plugins.go VolumePlugin interface
+// description and pkg/volume/util/volumehelper/volumehelper.go GetUniqueVolumeNameFromSpec that constructs
+// the unique volume names.
+func SplitUniqueName(uniqueName v1.UniqueVolumeName) (string, string, error) {
```
unit tests for this one?

kargakis commented on pkg/volume/util/volumehelper/volumehelper.go:
```diff
@@ -87,3 +88,18 @@ func GetUniqueVolumeNameFromSpec(
+func SplitUniqueName(uniqueName v1.UniqueVolumeName) (string, string, error) {
+	components := strings.SplitN(string(uniqueName), "/", 3)
+	if len(components) != 3 {
+		return "", "", fmt.Errorf("Cannot split volume unique name %s to plugin/volume components", uniqueName)
```
go errors by convention start with a lowercased letter

kargakis commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -255,25 +255,34 @@ func (asw *actualStateOfWorld) AddVolumeToReportAsAttached(
 func (asw *actualStateOfWorld) AddVolumeNode(
-	volumeSpec *volume.Spec, nodeName types.NodeName, devicePath string) (v1.UniqueVolumeName, error) {
+	uniqueName v1.UniqueVolumeName, volumeSpec *volume.Spec, nodeName types.NodeName, devicePath string) (v1.UniqueVolumeName, error) {
+	var volumeName v1.UniqueVolumeName
```
nit: move this below the locks

kargakis commented on pkg/controller/volume/attachdetach/attach_detach_controller_test.go:
```diff
@@ -47,3 +53,219 @@ func Test_NewAttachDetachController_Positive(t *testing.T) {
+		t.Fatalf("Fatal error encountered in the testing volume plugin")
+	}
+
+	close(stopCh)
```
You probably want to defer this

kargakis commented on pkg/controller/volume/attachdetach/attach_detach_controller_test.go:
```diff
@@ -47,3 +53,219 @@ func Test_NewAttachDetachController_Positive(t *testing.T) {
+	go adc.reconciler.Run(stopCh)
+	go adc.desiredStateOfWorldPopulator.Run(stopCh)
+
+	time.Sleep(time.Second * 1) // Wait so the reconciler calls sync at least once
```
this looks flaky

kargakis commented on pkg/controller/volume/attachdetach/attach_detach_controller_test.go:
```diff
@@ -47,3 +53,219 @@ func Test_NewAttachDetachController_Positive(t *testing.T) {
+		if i == 10 { // 10 seconds time out
+			t.Fatalf("Waiting for the volumes to attach/detach timed out: attached %d (expected %d); detached %d (%d)",
+				attachedVolumesNum, 1+extraPodsNum, detachedVolumesNum, nodesNum)
+			break
```
you don't need break when you fatalf

jsravn commented:
Shouldn't this change be using the shared informers so it uses the same store cache as everything else? Can't you just hook into the podUpdate/nodeUpdate methods? I'm not sure it's a great idea to list everything at startup here - it duplicates the api call (which may be very large in big clusters), and there is also a race with the informers.

jsravn commented:
I think the fundamental issue is actually https://github.com/kubernetes/kubernetes/issues/43300 - this is much more likely to cause the problem described than the edge case of missing a pod update. We need to fix the broken informer logic where there is a race between node informer and pod informer. Then to handle the case of missing a pod update during restart, we just need to update the nodeUpdate handler to parse the node information.

gnufied commented:
We definitely need to fix the race condition in pod & node informer discussed in issue #43300, but we need to fix this bug too.

It is pretty easy to reproduce for example:

1. Create pod with volumes in a multi-node environ. lets call the nodes N1 and N2.
2. Stop the controller 
3. delete the pods scheduled on N2
4. wait for pod resource to be removed from api server. it can take awhile, but it *will* get removed.
5. On N2, associated volume is unmounted and bind mount is removed but volume is still attached
6. Start the controller. The volume is still attached to the node and can't be reused.

jingxu97 commented:
@tsmetana what is the status of the PR? Are there still some tests failing?

jingxu97 commented:
Also I am thinking this PR might not solve the following issue

Because controller delete volume from  "VolumesAttached" list before actually performing detach operation, there is still a chance that we lose the information about the volume while it is still attached. The events might happen is 
1. pod is deleted 
2. controller checks the condition and remove volume from "VolumesAttached", but it has not yet called "detach" operation 
3. controller crash 
4. controller restarts, the volume information will not be found anywhere (neither from pod nor VolumesAttached in node status)

tsmetana commented:
> @tsmetana what is the status of the PR? Are there still some tests failing?

Sorry for the late response... The latest commit seems to be working OK. And yes, you're right -- there are more races affecting attach/detach and this patch fixes just one them.

tsmetana commented:
Rebased squashed, included fixes by @wongma7 (Thanks!).

wongma7 commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -65,7 +65,7 @@ type ActualStateOfWorld interface {
 	// If no node with the name nodeName exists in list of attached nodes for
 	// the specified volume, an error is returned.
-	SetVolumeMountedByNode(volumeName v1.UniqueVolumeName, nodeName types.NodeName, mounted bool) error
+	SetVolumeMountedByNode(volumeName v1.UniqueVolumeName, nodeName types.NodeName, mounted bool, forceUnmount bool) error
```
function's comment should be amended to include the new forceUnmount parameter. since the comment pretends atm that there's only one bool

saad-ali replied:
+1

wongma7 commented:
Overall the patch lgtm. The fishiest part is the use of SplitUniqueName during detach which relies on a kind of unenforced convention but we've already decided it's fine.

jing brings up a good issue, VolumesAttached is not as its name suggests the source of truth about what volumes are *actually* attached to the node. So there's still a hole there that needs plugging. IMO it could be fixed in a follow-up PR, this PR has gotten big enough already and fixes the more likely case where deletion is done while controller is down.

I imagine the fix for Jing's issue will have to either change the behavior around VolumesAttached or force us to introduce a new annotation "VolumesActuallyAttached" altogether, in which case we may as well not co-opt VolumesAttached and serialize the volume specs to the new annotation. But again, we can figure that stuff out in a follow-up.

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -239,12 +245,127 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+		nodeName := types.NodeName(node.Name)
+		for _, attachedVolume := range node.Status.VolumesAttached {
+			uniqueName := attachedVolume.Name
+			err = adc.actualStateOfWorld.MarkVolumeAsAttached(uniqueName, nil, nodeName, attachedVolume.DevicePath)
```
The `nil` value for `volumeSpec` scares me like @gnufied mentioned earlier. But it sounds like (https://github.com/kubernetes/kubernetes/pull/39732#discussion_r107663462) you are confident it won't break anything. 

nit: when passing in a value and it is unclear what the parameter is add the parameter name as a inline comment, e.g.:
``` GO
err = adc.actualStateOfWorld.MarkVolumeAsAttached(uniqueName, nil /* volumeSpec */, nodeName, attachedVolume.DevicePath)

```

nit: add a comment why you think it will be ok to pass in nil (e.g. https://github.com/kubernetes/kubernetes/pull/39732#discussion_r107663462).

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -239,12 +245,127 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+			uniqueName := attachedVolume.Name
+			err = adc.actualStateOfWorld.MarkVolumeAsAttached(uniqueName, nil, nodeName, attachedVolume.DevicePath)
+			if err != nil {
+				glog.Errorf("Error adding node to ActualStateOfWorld: %v", err)
```
Error seems not to reflect action.

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -239,12 +245,127 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+				continue
+			}
+			adc.processVolumesInUse(nodeName, node.Status.VolumesInUse, true /* forceUnmount */)
+			adc.desiredStateOfWorld.AddNode(types.NodeName(node.Name)) // Needed for DesiredStateOfWorld population
```
Check if the node is managed by attach/detach controller:
```GO
	if _, exists := node.Annotations[volumehelper.ControllerManagedAttachAnnotation]; exists {
		// Node specifies annotation indicating it should be managed by attach
		// detach controller. Add it to desired state of world.
		adc.desiredStateOfWorld.AddNode(nodeName)
	}
```

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -239,12 +245,127 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+				continue
+			}
+			nodeName := types.NodeName(podToAdd.Spec.NodeName)
+			plugin, err := adc.volumePluginMgr.FindPluginBySpec(volumeSpec)
```
`FindAttachablePluginBySpec` we don't care about non-attachable volumes in this controller.

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -239,12 +245,127 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+			}
+			nodeName := types.NodeName(podToAdd.Spec.NodeName)
+			plugin, err := adc.volumePluginMgr.FindPluginBySpec(volumeSpec)
+			if err != nil {
```
Check if the returned plugin is `nil`:
```GO
		if err != nil || attachableVolumePlugin == nil {
			glog.V(10).Infof(
				"Skipping volume %q for pod %q/%q: it does not implement attacher interface. err=%v",
				podVolume.Name,
				pod.Namespace,
				pod.Name,
				err)
			continue
		}

```

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -322,15 +443,15 @@ func (adc *attachDetachController) nodeDelete(obj interface{}) {
 // mounted.
 func (adc *attachDetachController) processVolumesInUse(
-	nodeName types.NodeName, volumesInUse []v1.UniqueVolumeName) {
+	nodeName types.NodeName, volumesInUse []v1.UniqueVolumeName, forceUnmount bool) {
```
Update comment to describe new boolean.

saad-ali reviewed (changes requested):
A couple of comments. Overall direction looks fine.

tsmetana commented:
Thanks for the comments. I hope I addressed them correctly in the new version (just pushed).

jsafrane commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -239,12 +245,136 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 		return
 	}
 
+	err := adc.populateActualStateOfWorld()
```
There is `controller.WaitForCacheSync()` just few lines above. Does it mean that all the existing objects were already synced through podAdd() and nodeAdd() callbacks from the shared informers? They may be synced in a wrong order (podAdd before nodeAdd) and these `populateActualStateOfWorld()` and `populateDesiredStateOfWorld()` will just fix it, right?

jsafrane replied:
I talked to @tsmetana and yes, these new world populators will fix states of worlds that could be inconsistent due to bad event ordering.

tsmetana replied:
The main issue is that we need to find volumes attached for the pods that were deleted during the controller downtime: this would not be solved by the WaitForCacheSync() -- we still would be missing the pod deletion event. And yes, in the case the events come in wrong order the ASW/DSW populators should help too.

jsafrane commented:
@k8s-bot cvm gce e2e test this
These are not attach/detach errors.

saad-ali commented:
Looks good, thanks for your patience @tsmetana!

/lgtm
/approve
